intercept htlcs

johncantrell97 opened the pull request:
This PR aims to add a new fake_scid Namespace (Namespace::Intercept) that allows htlcs forwarded with scids from this namespace to be surfaced to end-users via a new event (PaymentIntercepted).  The end-user must then either fail using `fail_intercepted_payment` or forward it using `forward_intercepted_payment` methods on ChannelManager.

The main use-case in mind is for enabling on-demand channel creation though there are likely other interesting use-cases that are also enabled by it.


I need help with updating serialization (read/write) for ChannelManager because we will need to add a new HashMap for storing the intercepted payments that we are waiting on the end-user to tell us what to do with them.  

I'm also not sure the best way to fail these.  Please see the note in `fail_intercepted_payment` about one thought I had.

I'm also not exactly sure what other functionality might be missing.  I added ability to generate a fake intercepter scid on the channel manager but there's nothing with route hints yet.  I'm not sure if it makes sense to add that to LDK or if it should be the end-user's responsibility because we can't be sure what they're doing with them? I guess they'll always at least want to include the fake_scid as the last hop but they'll likely want to specify a hop before that as well -- which in the on-demand channel / lsp use-case ldk won't know about this node.  

Lastly, I wrote about this on discord but I'm not sure what is needed in terms of signaling CM persistence when htlcs are drained from forward_htlcs => pending_intermittent_payments and then again moved back into foward_htlcs once signaled from user.  Feels like it's possible we lose an htlc in one of these transitions if the node is killed.

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3081,6 +3153,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+															new_events.push(events::Event::PaymentIntercepted {
+																short_channel_id: short_chan_id,
+																payment_hash,
+																inbound_amount_msats: 0,
```
I'm not sure best way to get this or if it's really that important to surface it? Does it need to be piped through from earlier steps?

johncantrell97 replied:
yeah that makes sense.  we can fail it safely because if it's missing that attribute it means it was created on a version that didn't support interception, right?

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3011,6 +3034,55 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+		if let Some(_payment) = pending_intercept {
+			// TODO: what's best way to fail this? in `process_pending_htlc_forwards` it uses that `fail_forward` macro
+			//			 awkward way could be to just use the forward_intercepted_payment code and pick a random scid
+		}
```
best way to handle this? see comment

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3011,6 +3034,55 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+
+							Ok(())
+					},
+					_ => Err(APIError::APIMisuseError { err: "impossible".to_string() })
```
Not sure how to handle this exactly.  It should be impossible to get into this state.

johncantrell97 replied:
with a) i assume when you say "store that struct here" you mean in pending_intercepted_payments?

b) does kind of suck but I guess would you accept a PR that just did that?

johncantrell97 replied:
went for a

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3081,6 +3153,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+															});
+														},
+														hash_map::Entry::Occupied(_) => {
+															fail_forward!(format!("Unknown short channel id {} for forward HTLC", short_chan_id), 0x4000 | 10, Vec::new(), None);
```
do we want to fail with this same error message in this case?

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1432,7 +1463,7 @@ macro_rules! handle_chan_restoration_locked {
 		let counterparty_node_id = $channel_entry.get().get_counterparty_node_id();
 		let res = loop {
-			let forwards: Vec<(PendingHTLCInfo, u64)> = $pending_forwards; // Force type-checking to resolve
+			let forwards: Vec<(PendingHTLCInfo, u64, Option<u64>)> = $pending_forwards; // Force type-checking to resolve
```
if i add it to PendingHTLCInfo then I wouldn't also have to thread through this extra Option<u64> in the tuple, right? 

It does seem to make sense to just put it in there.  it will still need to be an Option<u64> because they're serialized as part of pending forwards, right?

There's sort of another weirdness:

seems like in the Receive case, we are setting amt_to_forward to the 'prev/incoming amt' which ends up being fine because the logic differs for receives even though it's the same object.

i guess if i add this new field, would we want to update the receive logic as well to use it? though old PendingHtlcInfo's will still have the forward field set.  so we'd probably just want to store the incoming amount in both fields.. which is going to look pretty confusing.

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6696,6 +6833,12 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> Writeable f
+			intercept_id.write(writer)?;
+			pending_intercepted_payment.write(writer)?;
+		}
+
```
not sure where exactly this should be written. seemed like last was a good idea.

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6985,6 +7128,14 @@ impl<'a, Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
+				return Err(DecodeError::InvalidValue);
+			}
+		}
+
```
tried to read it last as well but the reader var gets moved into the read_tlv_fields! for outbound_payments below.  just put it here for now so it compiles but I'm pretty sure this won't work.

johncantrell97 commented:
ok moved the incoming amount to PendingHTLCInfo and now channel.rs is untouched.  I also attempted to update the ChannelManager read/write but I'm not entirely confident there.  I basically just added it to the end 🤷‍♂️ 

let me know what else might be missing when you get a chance.

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -131,6 +132,13 @@ pub(super) enum PendingHTLCStatus {
 	Fail(HTLCFailureMsg),
 }
 
+pub(super) struct PendingInterceptedHTLC {
```
Yeah, my bad.  agreed we should rename it if so.

johncantrell97 replied:
Looking at this now I'm a little confused.  Does HTLCForwardInfo::AddHTLC need to "hold" this struct? They are literally exactly the same already.  Was this not the case back in July when this was first written?

Unless I'm misremembering something couldn't I just get rid of this PendingInterceptedHTLC struct entirely and just store (or clone) the entire HTLCForwardInfo as the intercepted htlc instead of essentially copying it into this new struct?

johncantrell97 replied:
I guess it would mean matching on HTLCForwardInfo anywhere I need to use it and ignoring/erroring if we came across a ::FailHTLC when we expected an ::AddHTLC.

Do you know what you meant by hold it when it's already the same data?

johncantrell97 replied:
Oh whoops.  There's a separate comment thread in here about this.  You meant change it to HTLCForwardInfo::AddHTLC(RenamedPendingInterceptedHTLC).

Is the reason to update the AddHTLC variant to hold a struct instead of using named fields so that when I need to store the "pending intercepted htlcs" I can take (or clone) this inner struct instead of having to store the entire HTLCForwardInfo?  Thus also eliminating the need to match against it.

johncantrell97 replied:
So I starting working on this and it's a bit of a pain so I wanted to make sure it still makes sense to do this before I continue.

A couple things I wanted to highlight or questions I have about it if we do want to move forward with that idea:

1) There's a lot of places where you match on HTLCForwardInfo::AddHTLC { fields...} and do a nested match on the PendingRoutingHTLC::Forward variant at the same time.  I was doing this instead in the new version:

```
// destructuring to avoid having to write `pending_add_htlc_info.<field_name>` everywhere
let PendingAddHTLCInfo {
  prev_short_channel_id,
  prev_htlc_id,
  forward_info: PendingHTLCInfo {
    routing,
    incoming_shared_secret,
    payment_hash,
    amt_to_forward,
    outgoing_cltv_value,
    ..
  },
  prev_funding_outpoint
} = pending_add_htlc_info;

// instead of the nested match, putting the whole arm inside this if let instead :/
if let PendingHTLCRouting::Forward { onion_packet, .. } = routing {
  ...
}
```

Is there a cleaner way to do this?

2) How can I insert the PendingAddHTLCInfo (what I renamed it for now) into my map without cloning it? Is it possible? Is it okay if I have to clone it to insert it?  Otherwise it's fields get used / thus moved before I need to insert.

3) Is there an issue with serialization / backwards compat? The old enums will be serialized with the ::AddHTLC struct variant and the new version will want to deserialize with this tuple / nested struct version.  Is this okay?

johncantrell97 replied:
Ok sounds good, just wanted to make sure it made sense before continuing down this path. Thanks!

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3055,6 +3088,63 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+						PendingHTLCRouting::Forward { onion_packet, .. } => {
+							PendingHTLCRouting::Forward { onion_packet, short_channel_id: scid }
+						},
+						_ => payment.forward_info.routing
```
yeah, my gut is that we might as well support it. I assume this means surfacing the original onion in the event in addition to being able to pass a new one here, right?

johncantrell97 replied:
oh right, that makes sense. 

weren't there other non-lsp use-cases mentioned in here ( https://github.com/lightningdevkit/rust-lightning/issues/680 ) for a feature like this?

also it seems like this can still be useful for non-mpp supporting lsp.  i guess it really comes down to what trade-offs are being made with the mpp supporting lsp and if there's a strong case to support both flows.

johncantrell97 replied:
oh i guess i misunderstood 680 as well. i thought it alluded to being able surface htlcs to end-user similarly to this 'intercept' pr to enable them to add additional relay logic at the user-level.

johncantrell97 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6696,6 +6833,12 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> Writeable f
 			(11, self.probing_cookie_secret, required),
 		});
 
+		(pending_intercepted_payments.len() as u64).write(writer)?;
```
yeah that makes sense, whoops.

ariard reviewed (commented):
Could be nice to have a look on the pending or future spec objects which would require or be easier to implement with such HTLC intercept interface (e.g trampoline or https://github.com/lightning/bolts/issues/1008). To check if we fit them well.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3118,13 +3206,37 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 														},
 														_ => panic!(),
 													}
+												} else if forward_info.amt_incoming.is_some() && fake_scid::is_valid_intercept(&self.fake_scid_rand_bytes, short_chan_id) {
```
I thought of a issue potential attack vector here that I wanted to raise in the review club earlier today, but we didnät have time to go over the question about such vectors:

Shouldn't we check here that the `fake_scid` is a `scid` that we actually have generated, and not only that it's within the Intercept namespace?

Else, if another peer is able to generate a `fake_scid` that we determine to be in the Intercept namespace but that we didn't generate ourselves, we face 2 issues:

1. Since the `PaymentIntercepted` event is serialized for peers we have no active channel to (and isn't dropped on disconnect like other events for that type of peers), as well as `PendingInterceptedHTLC` which aren't handled, random peers we have no channels to could potentially be able cause extensive serialization of huge amount of events/structs.

2. Probably even worse, just one successful `fake_scid` attack by a random peer that passes the `is_valid_intercept`, will serialize the `PaymentIntercepted` event + `PendingInterceptedHTLC` for any LDK user, even users that didn't explicitly choose to use the new Intercept HTLC feature. That will break the backwards compatibility for such users, hence opening up for a backwards compatibility attack.

Looking at the implemenation for `fake_scids` + `namespaces`, it shouldn't be too expensive for random peers to brute force a single `fake_scid` that we will accept as a valid `Intercept` `fake_scid`? 

I think we therefore need to also add add a way to track which `fake_scids` we have actually generated to not face this issue? I guess one alternative solution to this would also be to add a config flag for enabling `Intercept HTLC `support, which defaults to `false`.

valentinewallace replied:
Good points, thanks!! I started a new discussion on the superceding PR here: https://github.com/lightningdevkit/rust-lightning/pull/1835#discussion_r1015967743. 

Re: (1) I'm not sure they could force us to store a ton of stuff since the payments would presumably be failed back as they come in by users that don't support intercept scid forwards (maybe the docs on the event should clarify that you MUST handle it)?

johncantrell97 commented:
> Let me know when you want additional rounds of review here!

Thanks, I just squashed all the old commits to ease the rebase.  I'm going to review what I was doing and the last round of comments here.  I think there's a few things that were never addressed.  

Any major changes in ldk that affect the ideas behind this change? During the rebase I noticed there's now a `forward_htlcs` map in channel manager that looks pretty similar to the pending_intercepted_payments I had added in this PR.

johncantrell97 commented:
> Let me know when you want additional rounds of review here!

This PR is kind of a mess with (outdated) comments already but would love feedback on these questions: https://github.com/lightningdevkit/rust-lightning/pull/1601/files#r987995094

moneyball commented:
Are we waiting on other contributors to respond to @johncantrell97's questions?

johncantrell97 commented:
> Are we waiting on other contributors to respond to @johncantrell97's questions?

Not at the moment. Think I could use a hint from matt about what existing routing code can be dry'd up now that I moved the new struct into the existing enum.

moneyball commented:
What's the next step here?

ariard commented:
> Could be nice to have a look on the pending or future spec objects which would require or be easier to implement with such HTLC intercept interface (e.g trampoline or https://github.com/lightning/bolts/issues/1008). To check if we fit them well.

I still think we should check if this proposed intercept HTLCs interface answers well the use-cases we have in mind such as trampoline/offline receive/jamming mitigation strategies and others. I still think there are previous review conceptual comments not addressed.

valentinewallace commented:
FYI @TheBlueMatt, as discussed offline with @johncantrell97, I'm almost done with a PR that supercedes this one (containing all of this PR's changes + tests)

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3055,6 +3088,63 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+					};
+
+					let mut per_source_pending_forward = vec![(payment.prev_short_channel_id, payment.prev_funding_outpoint, vec![(pending_htlc_info, payment.prev_htlc_id)])];
+					self.forward_htlcs(&mut per_source_pending_forward);
```
Added this in #1835, good catch IMO!

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5494,6 +5606,21 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+		loop {
+			let scid_candidate = fake_scid::Namespace::Intercept.get_fake_scid(best_block.height(), &self.genesis_hash, &self.fake_scid_rand_bytes, &self.keys_manager);
+			// Ensure the generated scid doesn't conflict with a real channel.
+			match channel_state.short_to_id.entry(scid_candidate) {
```
We only generate fake scids up to 144 blocks before our highest known height IIUC

valentinewallace commented:
Superceded by #1835. I think all the feedback here has been addressed but lmk if something's missing
